rust/src/treefile.rs

```diff
@@ -138,15 +140,21 @@ fn treefile_parse_stream<R: io::Read>(
     Ok(treefile)
 }
 
+/// Open file and provide context containing filename on failures.
+fn open_file<P: AsRef<Path>>(filename: P) -> Fallible<fs::File> {
+    return Ok(fs::File::open(filename.as_ref()).with_context(
+        |e| format!("Can't open file {:?}: {}", filename.as_ref().display(), e))?);
```

Comment:
@lucab was asking why we're doing `Ok(...?)` here instead of just `...`. I think what's happening there is that we're relying on the `.into()` behaviour of `?` in the case where opening the file fails. Doing just:

```rust
    return fs::File::open(filename.as_ref()).with_context(
        |e| format!("Can't open file {:?}: {}", filename.as_ref().display(), e));
```

fails with:

```rust
error[E0308]: mismatched types
   --> src/treefile.rs:145:12
    |
145 |       return fs::File::open(filename.as_ref()).with_context(
    |  ____________^
146 | |         |e| format!("Can't open file {:?}: {}", filename.as_ref().display(), e));
    | |________________________________________________________________________________^ expected struct `failure::Error`, found struct `failure::Context`
    |
    = note: expected type `std::result::Result<_, failure::Error>`
               found type `std::result::Result<_, failure::Context<std::string::String>>`
```

Similarly, doing `.into()` at the end will also fail because there's no conversion implementation from `Result<T, E>` to `Result<T, failure::Error>`. Whereas `failure::Context<T>` can be converted to `failure::Error` via the blanket `Fail` implementation.

Reply:
I feel like there's probably a `.map_err()` way to do this that would be more elegant but not worth redoing!